### ISSUE TITLE
Move to a jumpable random generator to avoid statistical depenency between map, flatMap, chain, etc calls

### ISSUE THREAD

**Issue by vlsi** [jlink/jqwik#423]
## Testing Problem

Currently, jqwik produces too many duplicate values for the input.

I believe the root cause is `nextSeed = random.nextLong()` pattern that is used in several key places.

For instance, below is a reproducer for `long nextSeed = random.nextLong();` in `net.jqwik.engine.properties.state.ShrinkableChain.ChainIterator#nextTransformer`: 

https://github.com/jlink/jqwik/blob/7a37d8dd5c66616320390e5c284a2e8eac6023fd/engine/src/main/java/net/jqwik/engine/properties/state/ShrinkableChain.java#L170-L172

The problem with `nextSeed = random.nextLong()` is that both random generators (the original and the one produced with `SourceOfRandomness.newRandom(nextSeed)`) would follow the same sequence.

For instance, try adding the following code to `ActionChainArbitraryTests`:

Frankly speaking, I find the syntax verbose, however, let it put aside for a while.

The idea of the test is that it creates `Set<Integer>` for the state, and performs one of the following actions:
1) clear the set
2) add a random integer between 1, Integer.MAX_VALUE there

Since I generate 30 actions, I would expect all the values to be unique

```java
	@Property(tries = 10000)
	void duplicatesInChain(@ForAll Random random) {
		ActionChainArbitrary<Set<Integer>> chains =
			ActionChain.<Set<Integer>>startWith(HashSet::new)
			   .addAction(
				   1, 
				   Action.<Set<Integer>>when(set -> !set.isEmpty())
					   .just(Transformer.mutate("clear", Set::clear))
			   )		
			   .addAction(
				   5, 
				   (Action.Dependent<Set<Integer>>) state ->
						Arbitraries.integers()
						   .between(1, 10000000)
						   .withDistribution(RandomDistribution.uniform())
						   .withoutEdgeCases()
						   .map(i -> {
									if (state.contains(i)) {
										// System.out.println("got duplicate value " + i);
										Statistics.label("value").collect("duplicate");
										return Transformer.noop();
									} else {
										Statistics.label("value").collect("unique");
										return Transformer.mutate("add " + i, set -> set.add(i));
									}
								}
						   )
			   )
			   .withMaxTransformations(30);

		ActionChain<Set<Integer>> chain = TestingSupport.generateFirst(chains, random);
		chain.run();
	}
```

The output is

```
[ActionChainArbitraryTests:duplicatesInChain] (530123) value = 
    unique    (510248) : 96 %
    duplicate ( 19875) :  4 %
```

If I replace `long nextSeed = random.nextLong()` with `long nextSeed = random.nextLong() + 1`, then I get much better results:

```
[ActionChainArbitraryTests:duplicatesInChain] (516820) value = 
    unique (516820) : 100 %
```

Even though using `+1` or another mixing function might be a good start, it would be great to use a jump algorithm with known good properties.

Just in case, here's a test that checks the behaviour of the regular `integers between` generator:

```java
	@Property(tries = 10000)
	void duplicatesInPlainArbitrary(@ForAll Random random) {
		Arbitrary<Integer> arbitrary =
			Arbitraries.integers()
					   .between(1, Integer.MAX_VALUE)
					   .withDistribution(RandomDistribution.uniform())
					   .withoutEdgeCases();
		Map<Integer, Long> count = TestingSupport.count(arbitrary.generator(30, false), 30, random);
		for (Map.Entry<Integer, Long> entry : count.entrySet()) {
			if (entry.getValue() == 1) {
				Statistics.label("value").collect("unique");
			} else {
				Statistics.label("value").collect("duplicate");
			}
		}
	}
```

```
[ActionChainArbitraryTests:duplicatesInPlainArbitrary] (300000) value = 
    unique (300000) : 100 %
```

## Suggested Solution

Move to a jumpable generator, and use `random.jump()` instead of `nextSeed = random.nextLong()`.

**Comment by jlink** [comment]
Reason may also be edge cases generation. Have you tried `@Property(edgeCases=NONE)`?

**Comment by vlsi** [comment]
Note that I explicitly configure `integers` into `.withoutEdgeCases()` mode, so I expect edge cases to have no impact:


Sample runs with `@Property(tries = 10000)`


```
    unique    (510464) : 97 %
    duplicate ( 18213) :  3 %

    unique    (510568) : 97 %
    duplicate ( 17941) :  3 %

    unique    (511124) : 97 %
    duplicate ( 17657) :  3 %
```

With `@Property(tries = 10000, edgeCases = EdgeCasesMode.NONE)`

```
    unique    (511608) : 97 %
    duplicate ( 18016) :  3 %

    unique    (510574) : 97 %
    duplicate ( 18235) :  3 %

    unique    (510822) : 97 %
    duplicate ( 17910) :  3 %
```

**Comment by vlsi** [comment]
I guess the most trivial solution here would be adding seed scramble to `XORShiftRandom(long seed)`.
Just in case, `new Random(long)` does scramble the given seed, and it does not use the seed "as is".

```diff
--- a/engine/src/main/java/net/jqwik/engine/SourceOfRandomness.java
+++ b/engine/src/main/java/net/jqwik/engine/SourceOfRandomness.java
@@ -65,7 +65,16 @@ public class SourceOfRandomness {
                        if (seed == 0l) {
                                throw new IllegalArgumentException("0L is not an allowed seed value");
                        }
-                       this.seed = seed;
+                       this.seed = mix64(seed);
+               }
+
+               /**
+                * See <a href="http://zimbry.blogspot.com/2011/09/better-bit-mixing-improving-on.html">Better Bit Mixing - Improving on MurmurHash3's 64-bit Finalizer</a>
+                */
+               private static long mix64(long z) {
+                       z = (z ^ (z >>> 30)) * 0xbf58476d1ce4e5b9L;
+                       z = (z ^ (z >>> 27)) * 0x94d049bb133111ebL;
+                       return z ^ (z >>> 31);
                }
```

yields 100% unique values when `.withDistribution(RandomDistribution.uniform()).withoutEdgeCases()` is used for generating integers.

`.withDistribution(RandomDistribution.uniform())` alone still yields 100% unique values

If I remove `uniform`, then I get

```
    unique    (517000) : 99.75 %
    duplicate (  1271) :  0.25 %
```

I guess it is expected to have some duplicates in case the generator is biased

`edgeCases = EdgeCasesMode.NONE` on `@Property` still makes no difference.

**Comment by vlsi** [comment]
Just in case, I indentified the issue when I analyzed the reasons my stateful chain produced lots of invalid (duplicate) actions.

**Comment by jlink** [comment]
Leaving the other discussion of having larger entropy in randomness aside, I see a much simpler solution:
I could just do a simple seed scrambling in `SourceOfRandomness.newRandom(long seed)` similar to what
`java.util.Random` does:

```java
private static long initialScramble(long seed) {
    return (seed ^ multiplier) & mask;
}
```

Probably without masking since `XORShiftRandom` actually uses 64bit of randomness and not just 48.

**Comment by jlink** [comment]
Just noticed that https://github.com/jlink/jqwik/pull/425 is doing mostly that.

**Comment by jlink** [comment]
I assume this can be closed with the PR #425 merged. Re-open if I'm wrong.

**Comment by vlsi** [comment]
Well, ideally, jqwik should use `random.split()` or `random.jump()` instead of `newRandom(random.nextLong())` when an independent generator is needed.

So I think this issue has not been resolved.
